python/dgl/nn/pytorch/explain/subgraphx.py

```diff
@@ -364,3 +365,445 @@ def explain_graph(self, graph, feat, target_class, **kwargs):
+        expanding children nodes from high degree to low degree when extending
+        the children nodes in the search tree. Otherwise, it will use the
+        "Low2high" strategy. Default: True
+    num_child : int, optional
+        This is the number of children nodes to expand when extending the
+        children nodes in the search tree. Default: 12
+    num_rollouts : int, optional
+        This is the number of rollouts for MCTS. Default: 20
+    node_min : int, optional
+        This is the threshold to define a leaf node based on the number of
+        nodes in a subgraph. Default: 3
+    shapley_steps : int, optional
+        This is the number of steps for Monte Carlo sampling in estimating
+        Shapley values. Default: 100
+    log : bool, optional
+        If True, it will log the progress. Default: False
+    """
+
+    def __init__(
+        self,
+        model,
+        num_hops,
+        coef=10.0,
+        high2low=True,
+        num_child=12,
+        num_rollouts=20,
+        node_min=3,
+        shapley_steps=100,
+        log=False,
+    ):
+        super().__init__()
+        self.num_hops = num_hops
+        self.coef = coef
+        self.high2low = high2low
+        self.num_child = num_child
+        self.num_rollouts = num_rollouts
+        self.node_min = node_min
+        self.shapley_steps = shapley_steps
+        self.log = log
+
+        self.model = model
+
+    def shapley(self, subgraph_nodes):
+        r"""Compute Shapley value with Monte Carlo approximation.
+
+        Parameters
+        ----------
+        subgraph_nodes : dict[str, Tensor]
+            subgraph_nodes[nty] gives the tensor node IDs of node type nty
+            in the subgraph, which are associated with this tree node
+
+        Returns
+        -------
+        float
+            Shapley value
+        """
+        num_nodes = self.graph.num_nodes()
+
+        # Obtain neighboring nodes of the subgraph g_i, P'.
+        local_regions = {
+            ntype: nodes.tolist() for ntype, nodes in subgraph_nodes.items()
+        }
+        for _ in range(self.num_hops - 1):
+            for c_etype in self.graph.canonical_etypes:
+                src_ntype, _, dst_ntype = c_etype
+                if (
+                    src_ntype not in local_regions
+                    or dst_ntype not in local_regions
+                ):
+                    continue
+
+                in_neighbors, _ = self.graph.in_edges(
+                    local_regions[dst_ntype], etype=c_etype
+                )
+                _, out_neighbors = self.graph.out_edges(
+                    local_regions[src_ntype], etype=c_etype
+                )
+                local_regions[src_ntype] = list(
+                    set(local_regions[src_ntype] + in_neighbors.tolist())
+                )
+                local_regions[dst_ntype] = list(
+                    set(local_regions[dst_ntype] + out_neighbors.tolist())
+                )
```

Comment:
I think L474-479 are problematic, as you can mix the node IDs of different node types together.

Reply:
what might an example of conditions for this? I figured mixing would not be possible when the type of `in_neighbors` are guarenteed to be `src_ntype` due to the `etype` scope when calling `graph.in_edges`? (and a similar argument for `out_neighbors`)

Reply:
I see. You are right. Thanks.